docs/docs/debugging-incremental-builds.md

```diff
@@ -0,0 +1,87 @@
+---
+title: Debugging Incremental Builds
+---
+
+With the [release of Gatsby v3](/docs/reference/release-notes/v3.0#incremental-builds-in-oss) Incremental Builds is available to everyone. This improvement is (re)generating only subset of HTML files that need to be generated. To be able to use it you will need to keep the `.cache` and `public` directories from previous builds.
+
+## How does it work?
+
+To be able to understand why you might see more generated pages than expected it's helpful to know how Gatsby tracks "inputs" when generating HTML files. In particular Gatsby tracks:
+
+- which page template the page is using
+- results of page query
+- results of static queries used by a page template
+- frontend source code (shared and also browser `gatsby-browser.js` / SSR `gatsby-ssr.js` specifically)
+
+When those inputs change since the last build, the HTML files are marked to be regenerated. If they don't change, Gatsby can reuse HTML files generated in previous build.
+
+## Debugging
+
+To be able to see the changed pages inside console, run `gatsby build` with some additional flags:
+
+```shell
+gatsby build --verbose --log-pages
+```
+
+At the end of the build you will see the changed pages:
+
+```shell
+info Built pages:
+Updated page: /foo/bar/
+```
+
+If you also want to see what and how files actually did change between builds you can look at the `diff` between two builds. Inside your project directory run these commands:
+
+```shell
+gatsby clean
+gatsby build --no-uglify
+cp -r public public-first-build
+```
+
+Now make your changes to the files and continue with these commands:
+
+```shell
+gatsby build --no-uglify --verbose --log-pages
+diff -u -r public-first-build public > build-diff.txt
```

Comment:
```suggestion
diff -u -r public-first-build public > build-diff.diff
```
Or something along those lines (I realize I used `.txt` in original comment, but using `.diff` would provide some syntax colouring in various dev tools)